File: Tennis.Logic/Game.cs
using System;

namespace Tennis.Logic
{
	public class Game
	{		
		private PointState sideOnePoints;
		private PointState sideTwoPoints;
		private GameState state;
		
		public Game()
		{
			this.sideOnePoints = PointState.Love;
			this.sideTwoPoints = PointState.Love;
			
			this.state = GameState.PriorToDeuce;
			
		}
		
		public void WinPoint(Func<Side, Side> scoring)
		{
			var scoringSide = scoring(Side.None);
			if (scoringSide == Side.One)
			{
				AdvanceState(scoringSide);
			}
			else if (scoringSide == Side.Two)
			{
				AdvanceState(scoringSide);
			}
		}
		
		public PointState SideOneScore
		{
			get { return sideOnePoints; }
		}
		
		public PointState SideTwoScore
		{
			get { return sideTwoPoints; }
		}
		
		public GameState State
		{
			get { return state; }
		}
		
		private void AdvanceState(Side side)
		{
			if (StillPlaying())
			{	
				if (HasEnteredDeuce())
				{
					this.state = GameState.Deuce;
					
					if (GetPointState(side) == PointState.Advantage)
					{
						this.state = DetermineWinner(side);
						return;
					}
					
					if((sideOnePoints == PointState.Deuce && sideTwoPoints == PointState.Deuce)
						|| (sideOnePoints == PointState.Forty && sideTwoPoints == PointState.Forty))
					{
						SetPointState(side, PointState.Advantage);
						return;
					}
					
					sideOnePoints = PointState.Deuce;
					sideTwoPoints = PointState.Deuce;
					return;
				}
				
				RollPoints(side);
				
				if (((int)GetPointState(side)) > 4)
				{
					this.state = DetermineWinner(side);
				}
			}
		}
		
		private bool StillPlaying()
		{
			return ((this.State == GameState.Deuce) || (this.State == GameState.PriorToDeuce));
		}
		
		private GameState DetermineWinner(Side side)
		{
			if (side == Side.One)
			{
				return GameState.GameWonBySideOne;
			}
			else
			{
				return GameState.GameWonBySideTwo;
			}
		}
		
		private void RollPoints(Side side)
		{
			var pointState = GetPointState(side);
			switch(pointState)
			{
				case PointState.None:
					SetPointState(side, PointState.Love);
					break;
				case PointState.Love:
					SetPointState(side, PointState.Fifteen);
					break;
				case PointState.Fifteen:
					SetPointState(side, PointState.Thirty);
					break;
				case PointState.Thirty:
					SetPointState(side, PointState.Forty);
					break;
				case PointState.Forty:
					SetPointState(side, PointState.Deuce);
					break;
			}
		}

		private bool HasEnteredDeuce ()
		{
			if (((int)sideOnePoints) >= 3 && ((int)sideTwoPoints) >= 3)
			{
				return true;
			}			
			return false;
		}
		
		public string PrintScore()
		{
			if (this.State != GameState.PriorToDeuce)
			{
				return TranslateDeuceOrBetterScore();
			}
			
			string sideOne = TranslateScore(this.sideOnePoints);
			string sideTwo = TranslateScore(this.sideTwoPoints);
			return sideOne + " - " + sideTwo;
		}
		
		private string TranslateDeuceOrBetterScore()
		{
			if (this.State == GameState.GameWonBySideOne)
			{
				return "game - side one";
			}
			else if(this.State == GameState.GameWonBySideTwo)
			{
				return "game - side two";
			}
			
			if (sideOnePoints == PointState.Advantage)
			{
				return "advantage - side one"; 
			}
			else if (sideTwoPoints == PointState.Advantage)
			{
				return "advantage - side two";
			}
			
			return "deuce";
		}
		
		private string TranslateScore(PointState pointState)
		{
			switch(pointState)
			{
			case PointState.Love:
				return "love";
			case PointState.Fifteen:
				return "fifteen";
			case PointState.Thirty:
				return "thirty";
			case PointState.Forty:
				return "forty";
			default:
			    return String.Empty;
			}
		}
		
		private PointState GetPointState(Side side)
		{
			if (side == Side.One)
			{
				return sideOnePoints;
			}
			else
			{
				return sideTwoPoints;
			}
		}
		
		private void SetPointState(Side side, PointState state)
		{
			if (side == Side.One)
			{
				sideOnePoints = state;
			}
			else
			{
				sideTwoPoints = state;
			}
		}
	}
}

File: Tennis.FSharp.Logic/Game.fs
namespace Tennis.FSharp.Logic

    open System
    open FSharpx

    type Game() =
        let mutable gameState = GameState.PriorToDeuce
        let mutable sideOnePointState = PointState.Love
        let mutable sideTwoPointState = PointState.Love

        let advancePointState sideWinnerPointState =
            match sideWinnerPointState with
            | PointState.Love -> PointState.Fifteen
            | PointState.Fifteen -> PointState.Thirty
            | PointState.Thirty -> PointState.Forty
            | PointState.Forty -> PointState.Deuce
            | PointState.Deuce -> PointState.Advantage
            | PointState.Advantage -> PointState.Win
            | _ -> PointState.Love

        let printScore pointState =
            match pointState with
            | PointState.Love -> "love"
            | PointState.Fifteen -> "fifteen"
            | PointState.Thirty -> "thirty"
            | PointState.Forty -> "forty"
            | PointState.Deuce -> "deuce"
            | PointState.Advantage -> "advantage"
            | _ -> "love"

        member private this.IsDeuce =
            if (int sideOnePointState) > 2 && (int sideTwoPointState) > 2 && sideOnePointState = sideTwoPointState then
                sideOnePointState <- PointState.Deuce
                sideTwoPointState <- PointState.Deuce
                gameState <- GameState.Deuce
                true
            else
                false

        member private this.IsAdvantageSideOne =
            if (int sideOnePointState) > 3 && (int sideTwoPointState) > 3 && sideOnePointState > sideTwoPointState && (int sideOnePointState) - (int sideTwoPointState) < 2 then
                sideOnePointState <- PointState.Advantage
                sideTwoPointState <- PointState.Deuce
                gameState <- GameState.Deuce
                true
            else
                false

        member private this.IsAdvantageSideTwo =
            if (int sideOnePointState) > 3 && (int sideTwoPointState) > 3 && sideTwoPointState > sideOnePointState && (int sideTwoPointState) - (int sideOnePointState) < 2 then
                sideOnePointState <- PointState.Deuce
                sideTwoPointState <- PointState.Advantage
                gameState <- GameState.Deuce
                true
            else
                false

        member private this.HasBeaten sideState sideToCompareState =
            if (int sideState) > 3 && sideState > sideToCompareState && (((int sideState) - (int sideToCompareState)) >= 2) then
                true
            else
                false
        
        member private this.DetermineWinner =
            if this.IsDeuce || this.IsAdvantageSideOne || this.IsAdvantageSideTwo then
                Side.None
            elif this.HasBeaten sideOnePointState sideTwoPointState then    
                gameState <- GameState.GameWonBySideOne
                Side.One
            elif this.HasBeaten sideTwoPointState sideOnePointState then
                gameState <- GameState.GameWonBySideTwo
                Side.Two
            else
                Side.None

        member this.SideOneScore
            with get () = sideOnePointState
            and set (value) = sideOnePointState <- value

        member this.SideTwoScore
            with get () = sideTwoPointState
            and set (value) = sideTwoPointState <- value

        member this.State
            with get () = gameState

        member this.PrintScore() =
            let winningSide = this.DetermineWinner

            if this.IsDeuce then
                "deuce"
            elif this.IsAdvantageSideOne then
                "advantage - side one"
            elif this.IsAdvantageSideTwo then
                "advantage - side two"
            elif winningSide = Side.None then
                let sideOneScore = printScore sideOnePointState
                let sideTwoScore = printScore sideTwoPointState
                sideOneScore + " - " + sideTwoScore
            else
                if winningSide = Side.One then
                    "game - side one"
                else
                    "game - side two"

        member this.WinPoint(f: Func<_,_>) =
            let func = FSharpFunc.FromFunc f
            let winner = func Side.None
            if winner = Side.One then
                sideOnePointState <- advancePointState sideOnePointState
                this.DetermineWinner
            else
                sideTwoPointState <- advancePointState sideTwoPointState
                this.DetermineWinner


